### PR TITLE
implement a method on the SendStream to send data immediately

### LINKED ISSUE #5575: non-blocking Write on Stream
Calls to `Stream.Write` might block if there’s insufficient (stream-level or connection-level) flow control credit available. This is the desired behavior in the vast majority of cases. 

However, in certain, very limited situations, it would be useful to do a non-blocking write on a stream. For example, WebTransport needs to send a WT_CLOSE_SESSION capsule when closing a session, and blocking this write call by not granting enough flow control credit is an attack vector.

The proposal is to add a non-blocking equivalent to Write, named `WriteImmediately([]byte) (int, error)`, to the `SendStream` and the `Stream`. This call will be guaranteed to not block under any circumstances. In cases where `Write` would have blocked, it would return a sentinel error. Therefore, it would also be unaffected by any write deadline set on the stream.

Question: What about partial writes, e.g. if there's only flow control window for a part of the message? The easiest solution seems writing whatever is allowed, and returning a `n > 0` with the sentinel error.

Note that `WriteImmediately` returning does not:
* guarantee that the data was sent out: the stream has a write buffer, and the Goroutine servicing the QUIC connection might not have packed data for this stream (and the connection could be closed before this happens, in which case the data would be discarded)
* take into account congestion control: similar to the previous point, congestion control is only taken into consideration when packets are packed, thus sending of the packet might be blocked by congestion control

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #5575.

> [!NOTE]
> ### Add `WriteImmediately` to `SendStream` for non-blocking writes
> - Adds `SendStream.WriteImmediately` (and `Stream.WriteImmediately`) that queues an entire byte slice without blocking on flow control or write deadlines. Returns `ErrWouldBlock` if there is insufficient flow control credit, with no partial queueing.
> - Introduces `TryAddBytesSent` on both stream and connection flow controllers to atomically reserve send credit at queue time, so later framing skips credit checks for reserved frames.
> - Refactors `baseFlowController` into a receive-only `receiveFlowController`, moving send-side state into dedicated fields with their own mutex on each controller for safe concurrent access.
> - `popNewOrRetransmittedStreamFrame` honors pre-reserved credit and emits `STREAM_DATA_BLOCKED` with the precise offset from the flow controller.
> - Risk: `CancelWrite` and `STOP_SENDING` handling now include reserved-but-unframed bytes in `FinalSize`; existing code that inspects final size may see larger values than before.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized d56f3ea.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added `WriteImmediately(p []byte)` for `Stream`/`SendStream` to queue data synchronously when send-window credit is available.
  * Added `ErrWouldBlock` to indicate the operation can’t be queued immediately due to flow control.
* **Bug Fixes**
  * Refined send-window credit handling and blocking detection.
  * Corrected reset/cancel and reliable-boundary sizing behavior when using `WriteImmediately`.
* **Tests**
  * Expanded coverage for `WriteImmediately`, flow-control blocking, frame splitting, and reset/reliable-boundary edge cases.

### CODE REVIEW

File: stream.go [hunk 1]
```diff
@@ -132,6 +132,12 @@ func (s *Stream) Write(p []byte) (int, error) {
 	return s.sendStr.Write(p)
 }
 
+// WriteImmediately writes data to the stream if it can be queued immediately.
+// See [SendStream.WriteImmediately] for more details.
+func (s *Stream) WriteImmediately(p []byte) error {
+	return s.sendStr.WriteImmediately(p)
+}
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply stream.go:139]: Looks like you're too stupid to read the issue.

File: send_stream.go [hunk 2]
```diff
@@ -103,6 +105,81 @@ func (s *SendStream) Write(p []byte) (int, error) {
...
+		return false, false, nil
+	}
+
+	bytesToReserve := protocol.ByteCount(len(p))
+	if s.nextFrame != nil && !s.nextFrameReserved {
+		bytesToReserve += s.nextFrame.DataLen()
+	}
+	if !s.flowController.TryAddBytesSent(bytesToReserve) {
+		return false, false, ErrWouldBlock
+	}
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply send_stream.go:154]: We don't refund anything, but we need to properly account for it when sending a RESET_STREAM frame.

[the remark this answers is not in the document]
Comment (marten-seemann) [reply send_stream.go:154]: We don't refund anything, but we need to properly account for it when sending a RESET_STREAM frame.

File: send_stream.go [hunk 3]
```diff
@@ -103,6 +105,81 @@ func (s *SendStream) Write(p []byte) (int, error) {
...
+
+	if s.nextFrame == nil {
+		s.nextFrame = wire.GetStreamFrame()
+		s.nextFrame.Offset = s.writeOffset
+		s.nextFrame.StreamID = s.streamID
+		s.nextFrame.DataLenPresent = true
+	}
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply send_stream.go:162]: This was fixed.

File: send_stream.go [hunk 4]
```diff
@@ -315,15 +411,24 @@ func (s *SendStream) popNewStreamFrame(maxBytes, maxDataLen protocol.ByteCount,
...
+				s.nextFrame = wire.GetStreamFrame()
+				s.nextFrame.Data = s.nextFrame.Data[:nextFrame.DataLen()-maxDataLen]
+			}
 			s.nextFrame.StreamID = s.streamID
 			s.nextFrame.Offset = s.writeOffset + maxDataLen
-			s.nextFrame.Data = s.nextFrame.Data[:nextFrame.DataLen()-maxDataLen]
 			s.nextFrame.DataLenPresent = true
 			copy(s.nextFrame.Data, nextFrame.Data[maxDataLen:])
 			nextFrame.Data = nextFrame.Data[:maxDataLen]
+			s.nextFrameReserved = nextFrameReserved
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply send_stream.go:425]: @macroscopeapp Is this still valid?

[the remark this answers is not in the document]
Comment (marten-seemann) [reply send_stream.go:425]: Yes, we never remove them, but we set a final offset that includes all reserved bytes, don't we? Please re-check.
